### ISSUE TITLE
Phalanx engine hangs PyChess on engine discovery

### ISSUE THREAD

**Issue by pychess** [pychess/pychess#535]
Original [issue 535](https://code.google.com/p/pychess/issues/detail?id=535) reported by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-20

**What steps will reproduce the problem?**
1. Install Phalanx
2. Start PyChess

**What is the expected output? What do you see instead?**

```
 Expected output is PyChess starting as normal with the addition of the
```

Phalanx engine.  Instead PyChess hangs on discovering the Phalanx engine.

**Does it happen every time?**

```
 Yes.
```

**What version of the product are you using?**

```
 r1635.
```

**Did you use an installed version of PyChess or did you run in from a**
**tarball/svn?**

```
 SVN
```

**Please provide any additional information below.**

Thanks for everything everyone!!! I love the new XDG specs -- good idea!!!
 Good luck with everything!  Thank you so much to everyone who helps make
PyChess my favorite Python Chess Program!  Thank yoU from the bottom of my
heart.  Thanks!

Sincerely,
John D. Zollo

**Please attach the latest pychess logfile. It's in a hidden folder, under**
your homedirectory, named ".local/share/pychess/"

**Comment by pychess** [comment]
- attached [2010-02-20_17-55-12.log](https://code.google.com/p/pychess/issues/detail?id=535#c0)

**Comment by pychess** [comment]
Original [comment 1.](https://code.google.com/p/pychess/issues/detail?id=535#c1) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-20

r1531 seems to be the last revision that worked correctly with Phalanx (and I was
able to play a game).  r1532 recognized the Phalanx engine, but when I tried to play,
I received errors (I've attached the log of r1532).

-John
- attached [2010-02-20_18-20-42.log](https://code.google.com/p/pychess/issues/detail?id=535#c1)

**Comment by pychess** [comment]
Original [comment 2.](https://code.google.com/p/pychess/issues/detail?id=535#c2) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-20

r1533 Also recognized Phalanx, but couldn't play a game.  (it would start, but black
would never make a move.)  r1534 through r1537, wouldn't run for me at all,
regardless of Phalanx (I believe because of a Python 2.5 bug), so I don't know if
Phalanx did or did not work in those revisions.  By the time we got to r1538, the
first bug I spoke of, hanging at discovery, appeared.  So, I don't know what happened
from r1531 that made it stop working, but it was working back there! (r1531 was the
last time Phalanx worked properly for me.)  My last comment mentions a different
Phalanx bug in r1532, but maybe it's unrelated to the engine discovery bug.  Who knows?

Thanks everyone!

-John

**Comment by pychess** [comment]
Original [comment 3.](https://code.google.com/p/pychess/issues/detail?id=535#c3) posted by [lobais](https://code.google.com/u/lobais/) 2010-02-21

I would like to install Phalanx myself for testing. Do you run Debian/Ubuntu, or do 
you know where to get it?

I appreciate your detailed descriptions. If possible I would like you to extend them 
with the following details:
- How do other cecp engines from http://code.google.com/p/pychess/wiki/ChessEngines 
  react?
- If you step back to r1533 does phalanx come to life again, or did it just work 
  back then?

Thanks, Thomas

**Comment by pychess** [comment]
Original [comment 4.](https://code.google.com/p/pychess/issues/detail?id=535#c4) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-21

Hey Thomas,

```
 I run Ubuntu 8.10 - Intrepid Ibex, so I just installed the Phalanx package
```

through Synaptic/Aptitude.  The project page for Phalanx is here:

http://phalanx.sourceforge.net/

The debian .debs are here:

http://packages.debian.org/search?keywords=phalanx&searchon=names&suite=all&section=all

The Ubuntu .debs are here:

http://packages.ubuntu.com/search?suite=default&section=all&arch=any&searchon=names&keywords=phalanx

Hopefully that will help you.

As far as stepping back, if I step back to r1531, Phalanx works fine.  So everything
with Phalanx worked fine in r1531.  If I step back to r1532, Phalanx gets recognized,
but when I start a game it doesn't move.  From r1534-r1537 I don't know what happens
with Phalanx cause there was a Python 2.5 bug and I run 2.5.  By r1538 (when the 2.5
bug was fixed) I receive the current hang, the same as the latest SVN revision.

As far as CECP: GnuChess, Gnome-GnuChess, and Crafty all work great!  The only CECP
engine I have a problem with is Phalanx.

Thanks Thomas!

Thank for all your help.  I sincerely appreciate it.

Thanks!
-John

**Comment by pychess** [comment]
Original [comment 5.](https://code.google.com/p/pychess/issues/detail?id=535#c5) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-22

I've been trying to figure out what happened between r1531 and r1532 that screwed
Phalanx.  The only error I'm getting is an assertion error (in r1532):

 File "/home/Z_John/PyChess/lib/pychess/Players/CECPEngine.py", line 266, in makeMove
    assert self.readyMoves
<type 'exceptions.AssertionError'> 

This only happens for Phalanx.  All the other CECP engines work fine in r1532, so I'm
just curious why only Phalanx had the problem with the switch from r1531 to r1532.
looking into it...

-John

**Comment by pychess** [comment]
Original [comment 6.](https://code.google.com/p/pychess/issues/detail?id=535#c6) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-22

Having the same problem with 2 CECP protover 1 engines:

From engineNest.py:

```
<engine protocol="cecp" protover="1" binname="phalanx">
```

and
    <engine protocol="cecp" protover="1" binname="boochess">

Both have the same assertion problem starting with r1532.

-John

**Comment by pychess** [comment]
Original [comment 7.](https://code.google.com/p/pychess/issues/detail?id=535#c7) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-22

Ok, so as far as that problem with r1532, I think I figured it out.

The only file that changed between r1531 and r1532 was CECPEngine.py
I think the section that got messed up was the definition of __startBlocking

I added back the 2 lines at Line 159, and it r1532 will work correctly with those
changes.  It looks like the change only broke protover 1 engines, while protover 2
engines worked correctly.  I added the line back, and phalanx and boochess work ok
with that custom hack.

I don't know enough about Python and PyChess to make a permamnent change, but that's
where the problem lies -- how to correctly fix it is up to you!

Sincerely,
john

**Comment by pychess** [comment]
Original [comment 8.](https://code.google.com/p/pychess/issues/detail?id=535#c8) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-22

Sorry about the tangent with CECPEngine.py -- althgouh it is a bug, it is not this
bug.  If you want I can start another issue -- up to you.  Going back the the
ORIGINAL bug:  First, one needs to make the changes I explained above to get CECP
protover engines (like Phalanx and BooChess) to work correctly in r1532 and r1533
(r1532 broke support for CECP protover 1 engines!!!).  After those changes have been
made (addressed above), one can see the REAL problem afterwords.  The real problem is
with engineNest.py.  Here's an explanation:

engineNest.py:

<= r1533  Everything works fine (except of course for the CECPEngine.py bug.)

r1534     engineNest.py now breaks Python 2.5.  Don't know if cecp protover 1 engines
work in 2.6

r1538     engineNest.py now works with Python 2.5, but CECP protover 1 engines don't
get discovered.  I don't know whether this bug appeared at r1534 or at r1538 because
I use Python 2.5 :(

Ok, that's it.  Those are the two problems:
1.  r1532 broke CECP protover 1 support in CECPEngine.py  (black would not move)
2.  Either r1534 or r1538 Broke CECP protover 1 support in engineNest.py  (Engines
   would not get discovered.)

Good luck!  Thanks for all your help -- I'm sorry I don't know Python well enough to
fix it myself, maybe some day.

Sincerely,
John

**Comment by pychess** [comment]
Original [comment 9.](https://code.google.com/p/pychess/issues/detail?id=535#c9) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-22

r1534 Caused this bug

r1538 is a minor tweak for Python 2.5.  The bug itself was introduced in r1534. 
Maybe someone with a newer version of Python can confirm this for me.

-John

**Comment by pychess** [comment]
Original [comment 10.](https://code.google.com/p/pychess/issues/detail?id=535#c10) posted by [lobais](https://code.google.com/u/lobais/) 2010-02-22

Hi John, thank you _sincerely_ for your deep work in unentangeling this bug!
I have looked through many of the revisions made, and while it is very likely that
r1534 messed something up, I don't see that r1532 could change anything for
protover=1. I can check it, when I get phalanx running, but until then, you might be
able to manually undo r1532 (I don't think it will create conflicts) and see if it
fixes the hang.

**Comment by pychess** [comment]
Original [comment 11.](https://code.google.com/p/pychess/issues/detail?id=535#c11) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-22

The hang started in r1534

**Comment by pychess** [comment]
Original [comment 12.](https://code.google.com/p/pychess/issues/detail?id=535#c12) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-22

nm... I'll just use XBoard in the meantime.

**Comment by pychess** [comment]
Original [comment 13.](https://code.google.com/p/pychess/issues/detail?id=535#c13) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-24

I think i figure out what the problem is.  Line 218 of engineNest.py (in r1534) is
calling a connect function that looks like it waits for: 'readyForOptions'.  Now,
'readyForOptions' works great with CECP protover 2 engines (as you can see in
CECPEngine.py on lines 712 and 714 (in r1534).  Unfortunately protover 1 CECP engines
never say "done".  I added another if statement for __parseLine in CECPEngine.py that
looks like this:

```
if "on" in parts:
    self.emit("readyForOptions")
    self.emit("readyForMoves")
    self.returnQueue.put("ready")
    return
```

This gets past the hangup on discovery.  I'm not sure if it works though -- one of
you may have to help me on it.

Thanks for the help!

Sincerely,
john

**Comment by pychess** [comment]
Original [comment 14.](https://code.google.com/p/pychess/issues/detail?id=535#c14) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-24

boochess is still broken.  I'm not sure how to parse the protover 1's -- any advice
would be greatly appreciated!

thanks!

-John

**Comment by pychess** [comment]
Original [comment 15.](https://code.google.com/p/pychess/issues/detail?id=535#c15) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-24

Ok, figured it out.  If you put this at the end of __parseline in CECPEngine.py:

```
if parts == ['xboard', 'mode', 'on']:
    #  Phalanx is ready
    self.emit("readyForOptions")
    self.emit("readyForMoves")
    self.returnQueue.put("ready")
    return

if parts == ['W', 'To', 'Move']:
    #  BooChess is ready
    self.emit("readyForOptions")
    self.emit("readyForMoves")
    self.returnQueue.put("ready")
    return
```

eveything works fine.  Check it out.  (I think you all could probably do it neater! 
I can't wait to see how you clean it up!)

thanks for everything.  Sincerely, thanks for being there.

Sincerely,
john

**Comment by pychess** [comment]
Original [comment 16.](https://code.google.com/p/pychess/issues/detail?id=535#c16) posted by [zollo.john](https://code.google.com/u/117741568078251192015/) 2010-02-24

This may not work with faile -- I don't have it, so I didn't try it.

-John

**Comment by pychess** [comment]
Original [comment 17.](https://code.google.com/p/pychess/issues/detail?id=535#c17) posted by [lobais](https://code.google.com/u/lobais/) 2010-02-24

Ah, I should have read that comment of yours before I started :)
You are right. The problem is, that in some revision of engineNest we decided to
using the readyForOptions signal as the signal to start reading parameters from the
engine and saving them to xml. In that process it was entirely forgotten that the
protover1 engines didn't ever send this signal. Thus in my version I changed the
'prestart' method to this:

```
def prestart (self):
    print >> self.engine, "xboard"
    if self.protover == 1:
        self.emit("readyForOptions")
    elif self.protover == 2:
        print >> self.engine, "protover 2"
        self.timeout = time.time() + TIME\_OUT\_FIRST
```

Adding the protover1 case.

This did however create another problem: The protover1 engines were finished too fast
for the discoverer dialog, that would get the all_engines_finished signal even before
the discovering_started signal.
So the latest rev contains a fix for this too.

Check it out,
and sorry for the long delays in communication.
Maybe you learned something useful about the codebase :) If there is any more details
you'd like me to explain, just write :)

Cheers
- changed status to: Fixed
